wip - scalacheck generators for blockchain + datastore

yusefnapora opened the pull request:
This adds a `genJournalBlock` function that creates scalacheck generator which returns a `JournalBlock` plus an in-memory "datastore" as a `Map[Reference, DataObject]`.  

You can feed these back into the `genJournalBlock` function, which will generate a new block + data objects, building on the previous one.

Needs some cleanup, and I want to test that the references are valid before using this for the client specs.

yusefnapora committed:
generate valid chain of cells for a CanonicalRecord

yusefnapora committed:
convenience constructor for MultihashReference for DataObject

yusefnapora committed:
add params to chain cell generators to set target canonical, chain

yusefnapora committed:
allow generating cells with nil chain references

yusefnapora committed:
factor out generating single cell for canonical record

yusefnapora committed:
allow passing in relationship References when generating cells

yusefnapora committed:
generate valid references between generated cells

yusefnapora committed:
generate one block at a time

yusefnapora committed:
don't "backtrack" through previous generated blocks

parkan commented:
I like having the generators in the protocol test namespace -- this gives us reusability (for example, the indexer client will probably need these) but without bloating distribution for protocol too much (should we distribute it)

parkan commented on protocol/src/test/scala/io/mediachain/protocol/CborSerializationSpec.scala:
```diff
@@ -217,10 +217,10 @@ object CborSerializationSpec extends BaseSpec with ScalaCheck {
-        genArtefactOwnershipCell,
-        genArtefactDerivationCell,
-        genArtefactCreationCell
+        genArtefactReferenceCell(),
```
do these have side effects now?

yusefnapora replied:
I made them into functions with default parameters, instead of just `val`s.  So that you can control what `entity` or `artefact` gets referenced by the cell, etc.

parkan commented on protocol/src/test/scala/io/mediachain/protocol/DataObjectGenerators.scala:
```diff
@@ -33,49 +39,87 @@ object DataObjectGenerators {
-  val genArtefactOwnershipCell = for {
-    base <- genArtefactChainCell
-    entity <- genReference
+  def genArtefactOwnershipCell(
```
I'm not sure we need these (i.e. specialized types beyond Update), though that may be a separate discussion. We're getting kind of close to the point where we need to have a serious think about how all this stuff is modeled.

parkan commented on protocol/src/test/scala/io/mediachain/protocol/JournalBlockGenerators.scala:
```diff
@@ -0,0 +1,195 @@
+    case a: ArtefactChainCell => a.artefact
+  }
+
+  def consChainCells(cells: List[ChainCell]): List[ChainCell] = {
```
can you add a small comment explaining what this does?

yusefnapora replied:
yeah, very good call :)

yusefnapora committed:
add `ref` field to `ChainCell` type

yusefnapora committed:
include generated block in mock datastore

yusefnapora committed:
only generate "update" cells, comments

yusefnapora committed:
add InMemoryDatastore that wraps a mutable Map[Reference, DataObject]

yusefnapora committed:
fix incorrect index for canonical entries

yusefnapora committed:
generator for blockchain with multiple blocks

yusefnapora committed:
add scalacheck Arbitrary instances for blockchain generators

yusefnapora committed:
use spec2-specific scalacheck parameters object

yusefnapora committed:
generate case classes instead of tuples to avoid implicit tuple madness

yusefnapora committed:
wip - client journal follower spec

yusefnapora commented:
@parkan I made a start on the spec for the client `JournalListener` implementation, but it's still pretty barebones.  I re-animated the `client` sub-project and put it in there, thinking it might be good to have common functionality like that in one place, and l-space and other future clients could depend on it...  what do you think; is that worth the extra project?

I still need to clean up the val vs def thing in the `DataObject` generators we were discussing yesterday, but got too focused on the blockchain generators this morning.  Will do that after lunch

parkan commented:
Yeah, the `client` core should be reusable for sure (for ex. for the indexer). This will also give us an easy way to migrate everything if/when we move off of copycat protocol.

yusefnapora committed:
check that JournalFollower finds all canonicals, chain cells referenced in blockchain

yusefnapora committed:
add client subproject to mediachain aggregate

yusefnapora committed:
add getAs[T] method to Datastore trait

yusefnapora committed:
don't keep all blocks in memory in JournalFollower, just keep refs

yusefnapora committed:
add spec for updating existing JournalFollower with new block

yusefnapora committed:
use sorted map for block references, sorted by journal index

- avoids redundant sorting when accessing the blocks
- avoids fetching all blocks from datastore just to get the current block
- also replaces mutable collections with vars of immutable collections

yusefnapora committed:
naive "folder" implementation for chain cells

Commit:
Rework metadata generators slightly

Commit:
Inter default generator states

Commit:
Rename generators for consistency

vyzo commented on client/src/main/scala/io/mediachain/client/CanonicalFolder.scala:
```diff
@@ -0,0 +1,97 @@
+        foldedArtefact(a, Some(c))
+
+      case _ =>
+        throw new IllegalStateException(
```
this is more of an IllegalArgumentException.

yusefnapora commented:
Going to close this and reopen with just the generators.
